x/interchainstaking/types/rebalance.go

```diff
@@ -57,140 +57,187 @@ func CalculateDeltas(currentAllocations map[string]sdkmath.Int, currentSum sdkma
+	currentLocked map[string]bool,
+	currentSum sdkmath.Int,
+	lockedSum sdkmath.Int,
+	targetAllocations ValidatorIntents,
+	logger log.Logger,
+) RebalanceTargets {
+	out := make(RebalanceTargets, 0)
+	targets, sources := CalculateDeltasNew(currentAllocations, currentLocked, currentSum, targetAllocations)
+
+	// rebalanceBudget = (total_delegations - locked)/2 == 50% of (total_delegations - locked)
+	// TODO: make this 2 (max_redelegation_factor) a param.
+	rebalanceBudget := currentSum.Sub(lockedSum).Quo(sdk.NewInt(2))
+
+	if logger != nil {
+		logger.Debug("Rebalancing", "total", currentSum, "totalLocked", lockedSum, "rebalanceBudget", rebalanceBudget)
+	}
+
+TARGET:
+	// targets are validators with a delegation deficit, sorted in descending order.
+	// that is, those at the top should be satisfied first to maximise progress toward goal.
+	for _, target := range targets {
+		// amount is amount we should try to redelegate toward target. This may be constrained by the remaining redelegateBudget.
+		// if it is zero (i.e. we hit the redelegation budget) break out of the loop.
+		amount := sdkmath.MinInt(target.Amount, rebalanceBudget)
+		if amount.IsZero() {
+			break
+		}
+		sources.Sort()
+		// sources are validators with available balance to redelegate, sorted in desc order.
+		for _, source := range sources {
+			switch {
+			case source.Amount.IsZero():
+				// if source is zero, skip.
+				continue
+			case source.Amount.GTE(amount):
+				// if source >= amount, fully satisfy target.
+				out = append(out, &RebalanceTarget{Amount: amount, Target: target.ValoperAddress, Source: source.ValoperAddress})
+				source.Amount = source.Amount.Sub(amount)
+				target.Amount = target.Amount.Sub(amount)
+				rebalanceBudget = rebalanceBudget.Sub(amount)
+				continue TARGET
+			case source.Amount.LT(amount):
+				// if source < amount, partially satisfy amount.
+				out = append(out, &RebalanceTarget{Amount: source.Amount, Target: target.ValoperAddress, Source: source.ValoperAddress})
+				amount = amount.Sub(source.Amount)
+				target.Amount = target.Amount.Sub(source.Amount)
+				rebalanceBudget = rebalanceBudget.Sub(source.Amount)
+				source.Amount = source.Amount.Sub(source.Amount)
+				if amount.IsZero() || rebalanceBudget.IsZero() {
+					// if the amount is fully satisfied or the rebalanceBudget is zero, skip to next target.
+					continue TARGET
+				}
+				// otherwise, try next source.
+			}
+		}
+		// we only get here if we are unable to satisfy targets due to rebalanceBudget depletion.
+		if logger != nil {
+			logger.Info("unable to satisfy targets with available sources.")
+		}
+	}
+
+	out.Sort()
 
 	return out
 }
 
+// func (d Deltas) Render() (out string) {
+// 	for _, delta := range d {
+// 		out = out + fmt.Sprintf("%s:\t%d\n", delta.ValoperAddress, delta.Amount.Int64())
+// 	}
+// 	return
+// }
+
 // MinDeltas returns the lowest value in a slice of Deltas.
 func MinDeltas(deltas ValidatorIntents) sdkmath.Int {
```

Comment:
```suggestion
func MinDelta(deltas ValidatorIntents) sdkmath.Int {
```

Should be singular since we are returning the one minimum delta value